server.py

```diff
@@ -16,159 +16,226 @@
+                else:
+                    return make_response("Product amount over restock_level", HTTP_400_BAD_REQUEST)
+
+            if info['type'] is 'new':
+                if total + info['quantity'] <= data['restock_level']:
+                    data = {'product_id': id, 'location_id': data['location_id'],
+                            'used': data['used'], 'new': data['new'] + info['quantity'],
+                            'open_box': data['open_box'], 'restock_level': data['restock_level']}
+                    inventory.put_product(id, data)
+                else:
+                    return make_response("Product amount over restock_level", HTTP_400_BAD_REQUEST)
+
+            if info['type'] is 'open_box':
+                if total + info['quantity'] <= data['restock_level']:
+                    data = {'product_id': id, 'location_id': data['location_id'],
+                            'used': data['used'], 'new': data['new'], 'open_box': data['open_box'] + info['quantity'],
+                            'restock_level': data['restock_level']}
+                    inventory.put_product(id, data)
+                else:
+                    return make_response("Product amount over restock_level", HTTP_400_BAD_REQUEST)
+
+        else:
+            make_response("Product data is not valid", HTTP_400_BAD_REQUEST)
+
+        return make_response(data.__str__(), HTTP_200_OK)
+
+    else:
+        return make_response("Product not found", HTTP_404_NOT_FOUND)
 
 
 @app.route('/inventory/products/<int:id>/remove', methods=['PUT'])
 def remove_from_product(id):
-  """ remove certain amount from product
+    """ remove certain amount from product
 
-  This method will remove certain amount from product
-  (eg. amount in new, open box or used.)
+    This method will remove certain amount from product
+    (eg. amount in new, open box or used.)
+
+    Args:
+      id (int): The id of the product to be removed from
 
-  Args:
-    id (int): The id of the product to be removed from
+    Returns:
+      response: remove successful message with status 200 if succeeded
+                or no product found with status 404 if cannot found the product
+                or invalid update with status 400 if the update violates any limitation.
 
-  Returns:
-    response: remove successful message with status 200 if succeeded
-              or no product found with status 404 if cannot found the product
-              or invalid update with status 400 if the update violates any limitation.
+    Todo:
+      * Finish the implementations.
+      * Write the tests for this.
 
-  Todo:
-    * Finish the implementations.
-    * Write the tests for this.
+    """
+    data = inventory.get_product(id)
+    if data is not None:
+        info = request.get_json()
+        if is_valid(info):
+            total = data['used'] + data['new'] + data['open_box']
+            if info['type'] is 'used':
+                if data['used'] - info['quantity'] >= 0:
```

Comment:
And I would also suggest using str constant USE, NEW, OPEN_BOX instead of 'used', 'new', in the Product.class

It's less error prone when you need it in multiple places.